engine/engine_test.go

```diff
@@ -347,6 +347,242 @@ func TestEngine_NotifyPutThenNotifyRemove(t *testing.T) {
+	provider2id := "Provider 2"
+	provider2Addrs := []string{"/ip4/0.0.0.0/tcp/4321/http"}
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID := []byte("fish")
+	subject.RegisterMultihashLister(func(ctx context.Context, p string, contextID []byte) (provider.MultihashIterator, error) {
+		if string(contextID) == string(wantContextID) {
+			return provider.SliceMultihashIterator(mhs), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, provider1id, provider1Addrs, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+	require.Equal(t, 1, subject.Chunker().Len())
+
+	gotPutAdCid2, err := subject.NotifyPut(ctx, provider2id, provider2Addrs, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid2)
+	require.Equal(t, 1, subject.Chunker().Len())
+}
+
+func TestEngine_DatastoreBackwardsCompativility_EmptyProvideEntriesShouldBeStoredWithoutPostfix(t *testing.T) {
+	ctx := contextWithTimeout(t)
+	rng := rand.New(rand.NewSource(1413))
+
+	mhs := testutil.RandomMultihashes(t, rng, 42)
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	providerId := "Provider 1"
+	providerAddrs := []string{"/ip4/0.0.0.0/tcp/4321/http"}
+
+	wantContextID := []byte("fish")
+	subject.RegisterMultihashLister(func(ctx context.Context, p string, contextID []byte) (provider.MultihashIterator, error) {
+		if string(contextID) == string(wantContextID) {
+			return provider.SliceMultihashIterator(mhs), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, "", []string{}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+	ad1, _ := subject.GetAdv(ctx, gotPutAdCid1)
+
+	gotPutAdCid2, err := subject.NotifyPut(ctx, providerId, providerAddrs, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid2)
+
+	ds := subject.Datastore()
+	// verifying that values for the default provider aren't postfixed with provider ID
+	entriesCid := ad1.Entries.(cidlink.Link).Cid.String()
+	val, _ := ds.Get(ctx, datastore.NewKey("map/cidKey/"+entriesCid))
+	require.Equal(t, string(wantContextID), string(val))
+	val, _ = ds.Get(ctx, datastore.NewKey("map/keyCid/"+string(wantContextID)))
+	_, gotCid, _ := cid.CidFromBytes(val)
+	require.Equal(t, entriesCid, gotCid.String())
+	has, _ := ds.Has(ctx, datastore.NewKey("map/cidKey/"+subject.ProviderID()+"/"+entriesCid))
+	require.False(t, has)
+	has, _ = ds.Has(ctx, datastore.NewKey("map/keyCid/"+subject.ProviderID()+"/"+string(wantContextID)))
+	require.False(t, has)
```

Comment:
Would be good to generate a datastore from the `main` branch, and check it into code as `testdata`. We can then write a test that asserts all exported methods in engine work correctly given the initial datastore _and_ publishing new ads on top of the old one works gracefully.

Please remember to copy the `testdata` datastore into a temp directory and use the temp copy for testing since engine modifies it.